// @microsoft/kiota-http-fetchlibrary/dist/es/src/middlewares/options/headersInspectionOptions.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
import { Headers } from "@microsoft/kiota-abstractions";
export const HeadersInspectionOptionsKey = "HeadersInspectionOptionsKey";
/**
 * RequestOption
 * Options
 * Options to inspect headers
 */
export class HeadersInspectionOptions {
    /**
     * Gets the request headers
     * @returns the request headers
     */
    getRequestHeaders() {
        return this.requestHeaders;
    }
    /**
     * Gets the response headers
     * @returns the response headers
     */
    getResponseHeaders() {
        return this.responseHeaders;
    }
    /**
     *
     * To create an instance of HeadersInspectionOptions
     * @param [options] - The headers inspection options value
     * @returns An instance of HeadersInspectionOptions
     * @example const options = new HeadersInspectionOptions({ inspectRequestHeaders: true, inspectResponseHeaders: true });
     */
    constructor(options = {}) {
        var _a, _b;
        this.requestHeaders = new Headers();
        this.responseHeaders = new Headers();
        this.inspectRequestHeaders = (_a = options.inspectRequestHeaders) !== null && _a !== void 0 ? _a : false;
        this.inspectResponseHeaders = (_b = options.inspectResponseHeaders) !== null && _b !== void 0 ? _b : false;
    }
    getKey() {
        return HeadersInspectionOptionsKey;
    }
}
